Importar liberias

In [1]:
import pandas as pd
import numpy as np


In [2]:
pip install dateparser

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.0.1 -> 25.3
[notice] To update, run: C:\Users\YD148ZK\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [3]:
import dateparser

abrir archivo CSV

In [4]:
pd.set_option('display.max_columns', None)
df_bank = pd.read_csv("../data/output/bank-additional.csv")
df_bank.head(5)

Unnamed: 0   age        job  marital    education  default  housing  loan  \
0           0   NaN  housemaid  MARRIED     basic.4y      0.0      0.0   0.0   
1           1  57.0   services  MARRIED  high.school      NaN      0.0   0.0   
2           2  37.0   services  MARRIED  high.school      0.0      1.0   0.0   
3           3  40.0     admin.  MARRIED     basic.6y      0.0      0.0   0.0   
4           4  56.0   services  MARRIED  high.school      0.0      0.0   1.0   

     contact  duration  campaign  pdays  previous     poutcome  emp.var.rate  \
0  telephone       261         1    999         0  NONEXISTENT           1.1   
1  telephone       149         1    999         0  NONEXISTENT           1.1   
2  telephone       226         1    999         0  NONEXISTENT           1.1   
3  telephone       151         1    999         0  NONEXISTENT           1.1   
4  telephone       307         1    999         0  NONEXISTENT           1.1   

  cons.price.idx cons.conf.idx euribor3m nr.employed   y                date  \
0         93,994         -36,4     4,857        5191  no       2-agosto-2019   
1         93,994         -36,4       NaN        5191  no  14-septiembre-2016   
2         93,994         -36,4     4,857        5191  no     15-febrero-2019   
3         93,994         -36,4       NaN        5191  no   29-noviembre-2015   
4         93,994         -36,4       NaN        5191  no       29-enero-2017   

   latitude  longitude                                   id_  
0    41.495    -71.233  089b39d8-e4d0-461b-87d4-814d71e0e079  
1    34.601    -83.923  e9d37224-cb6f-4942-98d7-46672963d097  
2    34.939    -94.847  3f9f49b5-e410-4948-bf6e-f9244f04918b  
3    49.041    -70.308  9991fafb-4447-451a-8be2-b0df6098d13e  
4    38.033   -104.463  eca60b76-70b6-4077-80ba-bc52e8ebb0eb

cambiar de "float" a "int" columnas : " age", "default", "housing", "loan".

In [5]:
df_bank.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43000 entries, 0 to 42999
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      43000 non-null  int64  
 1   age             37880 non-null  float64
 2   job             42655 non-null  object 
 3   marital         42915 non-null  object 
 4   education       41193 non-null  object 
 5   default         34019 non-null  float64
 6   housing         41974 non-null  float64
 7   loan            41974 non-null  float64
 8   contact         43000 non-null  object 
 9   duration        43000 non-null  int64  
 10  campaign        43000 non-null  int64  
 11  pdays           43000 non-null  int64  
 12  previous        43000 non-null  int64  
 13  poutcome        43000 non-null  object 
 14  emp.var.rate    43000 non-null  float64
 15  cons.price.idx  42529 non-null  object 
 16  cons.conf.idx   43000 non-null  object 
 17  euribor3m       33744 non-null 

In [6]:
df_bank.isnull().sum()

Unnamed: 0           0
age               5120
job                345
marital             85
education         1807
default           8981
housing           1026
loan              1026
contact              0
duration             0
campaign             0
pdays                0
previous             0
poutcome             0
emp.var.rate         0
cons.price.idx     471
cons.conf.idx        0
euribor3m         9256
nr.employed          0
y                    0
date               248
latitude             0
longitude            0
id_                  0
dtype: int64

In [7]:
df_bank.isnull().mean()*100

Unnamed: 0         0.000000
age               11.906977
job                0.802326
marital            0.197674
education          4.202326
default           20.886047
housing            2.386047
loan               2.386047
contact            0.000000
duration           0.000000
campaign           0.000000
pdays              0.000000
previous           0.000000
poutcome           0.000000
emp.var.rate       0.000000
cons.price.idx     1.095349
cons.conf.idx      0.000000
euribor3m         21.525581
nr.employed        0.000000
y                  0.000000
date               0.576744
latitude           0.000000
longitude          0.000000
id_                0.000000
dtype: float64

In [8]:
df_bank.duplicated().sum()

np.int64(0)

In [9]:
df_bank.describe().T

count          mean           std      min          25%  \
Unnamed: 0    43000.0  20651.099047  11868.770913    0.000  10381.75000   
age           37880.0     39.977112     10.437957   17.000     32.00000   
default       34019.0      0.000088      0.009390    0.000      0.00000   
housing       41974.0      0.535998      0.498708    0.000      0.00000   
loan          41974.0      0.155620      0.362499    0.000      0.00000   
duration      43000.0    257.739279    258.666033    0.000    102.00000   
campaign      43000.0      2.567233      2.772294    1.000      1.00000   
pdays         43000.0    962.330953    187.260394    0.000    999.00000   
previous      43000.0      0.174023      0.497366    0.000      0.00000   
emp.var.rate  43000.0      0.077128      1.573898   -3.400     -1.80000   
latitude      43000.0     36.856697      7.225948   24.396     30.61475   
longitude     43000.0    -95.939067     16.752282 -124.997   -110.49425   

                     50%          75%        max  
Unnamed: 0    20642.5000  30930.25000  41187.000  
age              38.0000     47.00000     98.000  
default           0.0000      0.00000      1.000  
housing           1.0000      1.00000      1.000  
loan              0.0000      0.00000      1.000  
duration        179.0000    319.00000   4918.000  
campaign          2.0000      3.00000     56.000  
pdays           999.0000    999.00000    999.000  
previous          0.0000      0.00000      7.000  
emp.var.rate      1.1000      1.40000      1.400  
latitude         36.7610     43.11325     49.384  
longitude       -95.8995    -81.42775    -66.937

In [10]:
df_bank["pdays"].unique()

array([999,   6,   4,   3,   5,   1,   0,  10,   7,   8,   9,  11,   2,
        12,  13,  14,  15,  16,  21,  17,  18,  22,  25,  26,  19,  27,
        20])

In [11]:
df_bank["pdays"].value_counts()

pdays
999    41412
3        457
6        428
4        121
9         67
7         66
2         63
12        61
10        55
5         50
13        38
11        30
15        26
1         26
14        22
8         20
0         15
16        13
17        10
18         7
22         4
19         3
21         2
25         1
26         1
27         1
20         1
Name: count, dtype: int64

Para la gran mayoria de los clientes siendo "pdays": Número de días que han pasado desde la última vez que se contactó con el cliente durante esta campaña. Vemos que en el mayor de los casos el resultado es 999, por lo que no se ha contactado recientemente con la mayoria de los clientes.

In [12]:
df_bank["previous"].unique()

array([0, 1, 2, 3, 4, 5, 6, 7])

In [13]:
df_bank["previous"].value_counts()

previous
0    37103
1     4778
2      786
3      232
4       75
5       20
6        5
7        1
Name: count, dtype: int64

Siendo la columna	previous: Número de veces que se ha contactado con el cliente antes de esta campaña. en la mayoria de los casos se ha contactado previamente "0" veces.

In [14]:
df_bank.select_dtypes(include="O").columns

Index(['job', 'marital', 'education', 'contact', 'poutcome', 'cons.price.idx',
       'cons.conf.idx', 'euribor3m', 'nr.employed', 'y', 'date', 'id_'],
      dtype='object')

In [15]:
df_bank.select_dtypes(include="int").columns

Index(['Unnamed: 0', 'duration', 'campaign', 'pdays', 'previous'], dtype='object')

In [16]:
df_bank.select_dtypes(include="float").columns

Index(['age', 'default', 'housing', 'loan', 'emp.var.rate', 'latitude',
       'longitude'],
      dtype='object')

cambiar de "float" a "int" columnas : " age", "default", "housing", "loan". Pero hay nulos en estas columnas por lo que no se puede aun modificar ./
y columna "date" tipo date - tambien tiene nulos - aunque lo primero que hay que hacer en esta columna será modificar los meses en español a ingles o numerico para que se pueda hacer el cambio a fecha

In [17]:
df_bank.columns

Index(['Unnamed: 0', 'age', 'job', 'marital', 'education', 'default',
       'housing', 'loan', 'contact', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp.var.rate', 'cons.price.idx',
       'cons.conf.idx', 'euribor3m', 'nr.employed', 'y', 'date', 'latitude',
       'longitude', 'id_'],
      dtype='object')

In [18]:

df_bank.rename(columns={"id_": "ID"}, inplace=True)


In [19]:
df_bank.columns

Index(['Unnamed: 0', 'age', 'job', 'marital', 'education', 'default',
       'housing', 'loan', 'contact', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp.var.rate', 'cons.price.idx',
       'cons.conf.idx', 'euribor3m', 'nr.employed', 'y', 'date', 'latitude',
       'longitude', 'ID'],
      dtype='object')

In [20]:
'''float_int = ['age', 'default','housing', 'loan']
for col in float_int:
    df_bank[col] = df_bank[col].apply(lambda x: int(x))'''

"float_int = ['age', 'default','housing', 'loan']\nfor col in float_int:\n    df_bank[col] = df_bank[col].apply(lambda x: int(x))"

In [21]:
for col in df_bank.select_dtypes(include="O").columns:
    df_bank[col] = df_bank[col].str.lower()

df_bank.tail(5)

Unnamed: 0   age          job  marital            education  default  \
42995       19154   NaN       admin.  married    university.degree      0.0   
42996       26206  34.0   technician  married  professional.course      0.0   
42997       15046   NaN  blue-collar   single             basic.6y      0.0   
42998       15280   NaN       admin.  married    university.degree      NaN   
42999       27570   NaN   unemployed   single    university.degree      0.0   

       housing  loan   contact  duration  campaign  pdays  previous  \
42995      0.0   0.0  cellular       618         2    999         0   
42996      1.0   1.0  cellular        42         7    999         0   
42997      1.0   0.0  cellular       391         2    999         0   
42998      0.0   0.0  cellular       674         3    999         0   
42999      0.0   1.0  cellular       104         2    999         0   

          poutcome  emp.var.rate cons.price.idx cons.conf.idx euribor3m  \
42995  nonexistent           1.4         93,444         -36,1       NaN   
42996  nonexistent          -0.1           93,2           -42       NaN   
42997  nonexistent           1.4         93,918         -42,7       NaN   
42998  nonexistent           1.4         93,918         -42,7     4,958   
42999  nonexistent          -0.1           93,2           -42     4,021   

      nr.employed    y                date  latitude  longitude  \
42995      5228,1  yes     13-octubre-2015    38.147   -105.582   
42996      5195,8   no       17-marzo-2018    49.235   -112.201   
42997      5228,1   no  15-septiembre-2016    40.679   -120.015   
42998      5228,1   no  23-septiembre-2019    27.772   -117.518   
42999      5195,8   no    6-noviembre-2019    41.146   -105.026   

                                         ID  
42995  4eed05de-2a98-4227-b488-32122009b638  
42996  0f0aca88-4088-4fe2-905f-44fb675d9493  
42997  cadadd4b-7ee5-4019-b13a-ca01bb67ca5b  
42998  5f432048-d515-4bb5-9c94-62db451f88d4  
42999  993bbbd6-4dbc-4a40-a408-f91f8462bee6

In [22]:

for col in df_bank.select_dtypes(include="object").columns:
        print(col.upper())
        print(df_bank[col].value_counts())    


JOB
job
admin.           10873
blue-collar       9654
technician        7026
services          4162
management        3050
retired           1790
entrepreneur      1522
self-employed     1489
housemaid         1123
unemployed        1063
student            903
Name: count, dtype: int64
MARITAL
marital
married     25999
single      12105
divorced     4811
Name: count, dtype: int64
EDUCATION
education
university.degree      12722
high.school             9925
basic.9y                6309
professional.course     5477
basic.4y                4356
basic.6y                2386
illiterate                18
Name: count, dtype: int64
CONTACT
contact
cellular     27396
telephone    15604
Name: count, dtype: int64
POUTCOME
poutcome
nonexistent    37103
failure         4461
success         1436
Name: count, dtype: int64
CONS.PRICE.IDX
cons.price.idx
93,994    7938
93,918    6937
92,893    5985
93,444    5349
94,465    4522
93,2      3731
93,075    2552
92,201     803
92,963     742
92,431     471
9

In [24]:
df_bank.to_csv("../data/output/bank-additional_01.csv", index=False)